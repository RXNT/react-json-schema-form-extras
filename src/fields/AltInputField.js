import React from "react";

function altInputField(props) {
  let {
    uiSchema: { defInput, altInput, altInputSeparator = "" },
    registry: { fields },
  } = props;

  let DefInput = fields[defInput];
  let AltInput = fields[altInput];

  let noBorder = {
    border: "0",
    boxShadow: "none",
  };
  return (
    <div className="row">
      <div className="col-md-5">
        <DefInput {...props} />
      </div>
      <div className="col-md-2 text-center">
        <div className="form-control" style={noBorder}>
          {altInputSeparator}
        </div>
      </div>
      <div className="col-md-5">
        <AltInput {...props} />
      </div>
    </div>
  );
}

export default altInputField;
