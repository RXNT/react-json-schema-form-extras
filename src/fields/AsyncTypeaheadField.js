import React, { Component } from "react";
import PropTypes from "prop-types";
import { AsyncTypeahead } from "react-bootstrap-typeahead";
import { search } from "./AsyncTypeaheadAPI";
import { isDevelopment, mapLabelKey, mapSchema } from "./utils";

const DEFAULT_OPTIONS = {
  required: false,
  labelKey: "name",
  minLength: 3,
  placeholder: "Search...",
};

class AsyncTypeaheadField extends Component {
  constructor(props) {
    super(props);

    this.state = {
      options: [],
    };
  }

  handleSearch = query => {
    if (!query) {
      return;
    }

    let { uiSchema: { typeahead: { url, optionsMapping } } } = this.props;

    search(url, query, optionsMapping).then(options =>
      this.setState({ options })
    );
  };

  handleSelectionChange = events => {
    if (events.length > 0) {
      let {
        schema,
        uiSchema: { typeahead: { responseSchemaMapping } },
        onChange,
      } = this.props;
      let schemaEvents = mapSchema(events, schema, responseSchemaMapping);
      onChange(schemaEvents);
      setTimeout(() => {
        if (this.refs.typeahead) {
          this.refs.typeahead.getInstance().clear();
        }
      }, 0);
    }
  };

  render() {
    let { uiSchema: { typeahead } } = this.props;

    let typeConf = Object.assign({}, DEFAULT_OPTIONS, typeahead);
    typeConf.onChange = this.handleSelectionChange;
    typeConf.onSearch = this.handleSearch;
    typeConf.options = this.state.options;
    typeConf.ref = "typeahead";
    typeConf.labelKey = mapLabelKey(typeahead.labelKey);

    return <AsyncTypeahead {...typeConf} />;
  }
}

if (isDevelopment()) {
  AsyncTypeaheadField.propTypes = {
    schema: PropTypes.object.isRequired,
    uiSchema: PropTypes.shape({
      typeahead: PropTypes.shape({
        url: PropTypes.string.required,
        optionsMapping: PropTypes.string,
        responseSchemaMapping: PropTypes.object,
      }),
    }),
  };
}

export default AsyncTypeaheadField;
