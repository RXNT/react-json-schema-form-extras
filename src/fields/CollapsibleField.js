import React from "react";

function CollapseMenu(props) {
  let { onChange, schema: { title }, collapsed } = props;
  let {
    uiSchema: {
      collapse: {
        icon: {
          enabled = "glyphicon glyphicon-chevron-down",
          disabled = "glyphicon glyphicon-chevron-right",
        } = {},
      },
      separate = true,
      wrapClassName = "lead",
    },
  } = props;
  return (
    <div className={wrapClassName}>
      <span>{title}</span>&nbsp;
      <a onClick={onChange}>
        <i className={collapsed ? disabled : enabled} />
      </a>
      {separate && <hr />}
    </div>
  );
}

class CollapsibleField extends React.Component {
  constructor(props) {
    super(props);

    let { uiSchema: { collapse: { active = true } = {} } } = props;

    this.state = { collapsed: active };
  }

  handleCollapsed = () => {
    this.setState(function(state) {
      return { collapsed: !state.collapsed };
    });
  };

  render() {
    let { schema, uiSchema, registry: { fields } } = this.props;
    let { collapsed } = this.state;
    let { collapse: { field } } = uiSchema;
    let CollapseElement = fields[field];

    return (
      <div>
        <CollapseMenu
          schema={schema}
          uiSchema={uiSchema}
          collapsed={collapsed}
          onChange={this.handleCollapsed}
        />
        <div className="form-group">
          {!collapsed && <CollapseElement {...this.props} />}
        </div>
      </div>
    );
  }
}

export default CollapsibleField;
