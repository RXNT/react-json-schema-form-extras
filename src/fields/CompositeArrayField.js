import React from "react";

class CompositeArrayField extends React.Component {
  handleAdd = list => {
    let { formData = [] } = this.props;
    let newTable = formData.concat(list);
    this.props.onChange(newTable);
  };

  render() {
    let {
      uiSchema: { inputField, arrayField },
      registry: { fields },
    } = this.props;

    let InputElement = fields[inputField];
    let ArrayElement = fields[arrayField];
    return (
      <div>
        <div className="form-group">
          <InputElement {...this.props} onChange={this.handleAdd} />
        </div>
        <div className="form-group">
          <ArrayElement {...this.props} />
        </div>
      </div>
    );
  }
}

export default CompositeArrayField;
