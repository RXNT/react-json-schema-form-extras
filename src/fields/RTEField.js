import React, { Component } from "react";
import RichTextEditor from "react-rte";

export default class RTEField extends Component {
  constructor(props) {
    super(props);

    let { formData = "" } = props;

    this.state = {
      value: RichTextEditor.createValueFromString(formData, "html"),
    };
  }

  handleChange = value => {
    this.setState({ value });
    if (this.props.onChange) {
      // Send the changes up to the parent component as an HTML string.
      // This is here to demonstrate using `.toString()` but in a real app it
      // would be better to avoid generating a string on each change.
      this.props.onChange(value.toString("html"));
    }
  };

  render() {
    let { uiSchema: { rte } } = this.props;

    return (
      <RichTextEditor
        {...rte}
        value={this.state.value}
        onChange={this.handleChange}
      />
    );
  }
}
