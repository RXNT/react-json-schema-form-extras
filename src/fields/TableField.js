import React, { Component } from "react";
import { BootstrapTable, TableHeaderColumn } from "react-bootstrap-table";

const DEFAULT_TABLE_CONF = {
  cellEdit: {
    mode: "click",
    blurToSave: true,
  },
  options: {},
  selectRow: {
    mode: "checkbox",
  },
  deleteRow: true,
  handleConfirmDeleteRow: next => next(),
};

const POSITION_KEY = "_position";

function addIndex(data) {
  return data.map((el, i) => Object.assign({}, el, { [POSITION_KEY]: i }));
}

function removeIndex(data) {
  return data.map(el => {
    let newEl = Object.assign({}, el);
    delete newEl[POSITION_KEY];
    return newEl;
  });
}

function toDataFormat(fieldProp) {
  if (fieldProp.enum && fieldProp.enumNames) {
    return cell => {
      return fieldProp.enumNames[fieldProp.enum.indexOf(cell)];
    };
  }
  return undefined;
}

function toEditable(fieldProp) {
  if (fieldProp.enum) {
    if (fieldProp.enumNames) {
      let values = fieldProp.enum.map((value, i) => {
        let text = fieldProp.enumNames[i];
        return { value, text };
      });
      return {
        type: "select",
        options: { values },
      };
    } else {
      return {
        type: "select",
        options: { values: fieldProp.enum },
      };
    }
  } else if (fieldProp.type === "boolean") {
    return {
      type: "checkbox",
    };
  } else if (fieldProp.type === "date-time") {
    return {
      type: "datetime",
    };
  } else if (fieldProp.type === "date") {
    return {
      type: "date",
    };
  } else if (fieldProp.type === "time") {
    return {
      type: "time",
    };
  }
  return true;
}

function setColumnCSSIfMissing(col, css) {
  col.className = col.className ? col.className : css;
  col.columnClassName = col.columnClassName ? col.columnClassName : css;
}

function withColumnCss(columns) {
  let shownColumns = columns.filter(
    col => col.hidden === undefined || !col.hidden
  );
  let numCols = shownColumns.length;
  let colSize = Math.round(12 / numCols);
  if (colSize == 0) {
    return columns;
  }

  let colCss = `col-md-${colSize}`;
  shownColumns.forEach((col, i) => {
    if (i != 0) {
      setColumnCSSIfMissing(col, colCss);
    }
  });
  return columns;
}

export function toTableColumns(schema, tableCols = []) {
  let { items: { properties } } = schema;

  let schemaCols = Object.keys(properties).map(dataField => {
    let { title } = properties[dataField];
    let editable = toEditable(properties[dataField]);
    let dataFormat = toDataFormat(properties[dataField]);
    return { dataField, displayName: title, editable, dataFormat };
  });

  let columnsWithOverrides = schemaCols.map(sCol => {
    let tCol = tableCols.find(col => col.dataField === sCol.dataField);
    return Object.assign(sCol, tCol);
  });

  return withColumnCss(columnsWithOverrides);
}

class TableField extends Component {
  toTableConf = () => {
    let { uiSchema: { table = {} }, formData = [] } = this.props;

    if (table.keyField === undefined || table.keyField === POSITION_KEY) {
      table.keyField = POSITION_KEY;
      formData = addIndex(formData);
    }

    let tableConf = Object.assign({}, DEFAULT_TABLE_CONF, table, {
      data: formData,
    });

    tableConf.cellEdit.afterSaveCell = this.handleCellSave;
    tableConf.options.afterDeleteRow = this.handleRowsDelete;

    return tableConf;
  };

  handleCellSave = updRow => {
    let { keyField, data } = this.toTableConf();

    const targetKey = updRow[keyField];
    let updTable = data.map(
      row => (row[keyField] === targetKey ? updRow : row)
    );

    this.props.onChange(removeIndex(updTable));
    return updTable;
  };

  handleRowsDelete = removedKeys => {
    const { keyField, data } = this.toTableConf();

    let filteredRows = data.filter(row => {
      let rowKey = row[keyField];
      return !removedKeys.includes(rowKey);
    });

    this.props.onChange(removeIndex(filteredRows));
  };

  render() {
    let tableConf = this.toTableConf();
    let columns = toTableColumns(this.props.schema, tableConf.tableCols);

    return (
      <BootstrapTable {...tableConf}>
        {columns.map((column, i) => {
          return (
            <TableHeaderColumn key={i} {...column}>
              {column.displayName}
            </TableHeaderColumn>
          );
        })}
      </BootstrapTable>
    );
  }
}

export default TableField;
