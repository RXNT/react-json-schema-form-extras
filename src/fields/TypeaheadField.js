import React, { Component } from "react";
import PropTypes from "prop-types";
import { Typeahead } from "react-bootstrap-typeahead";
import { mapLabelKey, mapSchema } from "./utils";

const DEFAULT_OPTIONS = {
  required: false,
  labelKey: "name",
  minLength: 3,
  placeholder: "Search...",
};

class TypeaheadField extends Component {
  handleSelectionChange = events => {
    if (events.length > 0) {
      let {
        schema,
        uiSchema: { typeaheadOptions: { mapping, cleanAfterSelection = true } },
      } = this.props;
      let schemaEvents = mapSchema(events, schema, mapping);
      this.props.onChange(schemaEvents);
      if (cleanAfterSelection) {
        setTimeout(() => {
          if (this.refs.typeahead) {
            this.refs.typeahead.getInstance().clear();
          }
        }, 0);
      }
    }
  };

  render() {
    let { uiSchema: { typeaheadOptions } } = this.props;

    let typeConf = Object.assign({}, DEFAULT_OPTIONS, typeaheadOptions);
    typeConf.onChange = this.handleSelectionChange;
    typeConf.labelKey = mapLabelKey(typeaheadOptions.labelKey);
    typeConf.ref = "typeahead";

    return <Typeahead {...typeConf} />;
  }
}

TypeaheadField.propTypes = {
  schema: PropTypes.object.isRequired,
  uiSchema: PropTypes.shape({
    genTypeahead: PropTypes.shape({
      options: PropTypes.array.required,
    }),
  }),
};

export default TypeaheadField;
