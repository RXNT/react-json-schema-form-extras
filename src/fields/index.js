import TableField from "./TableField";
import AsyncTypeaheadField from "./AsyncTypeaheadField";
import TypeaheadField from "./TypeaheadField";
import CompositeArrayField from "./CompositeArrayField";
import AltInputField from "./AltInputField";
import CollapsibleField from "./CollapsibleField";
import IMOTypeaheadField from "./imo/IMOField";
import RTEField from "./RTEField";

export default {
  table: TableField,
  asyncTypeahead: AsyncTypeaheadField,
  typeahead: TypeaheadField,
  collapsible: CollapsibleField,
  compositeArray: CompositeArrayField,
  altInput: AltInputField,
  rte: RTEField,
  imo: IMOTypeaheadField,
};
