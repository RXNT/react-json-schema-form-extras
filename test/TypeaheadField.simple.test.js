import React from "react";
import Form from "react-jsonschema-form";
import fields from "../src";
import renderer from "react-test-renderer";

let formData = {
  str: "some",
  strArr: ["some"],
  obj: { name: "other" },
  objArr: [{ name: "other" }],
};

let schema = {
  type: "object",
  properties: {
    str: {
      type: "string",
    },
    strArr: {
      type: "array",
      items: {
        type: "string",
      },
    },
    obj: {
      type: "object",
      properties: {
        name: { type: "string" },
      },
    },
    objArr: {
      type: "array",
      items: {
        type: "object",
        properties: {
          name: { type: "string" },
        },
      },
    },
  },
};

let fieldProps = {
  classNames: "col-md-12",
  "ui:field": "typeahead",
  typeahead: {
    options: [{ name: "some" }, { name: "other" }],
    labelKey: "name",
    minLength: 1,
  },
};

let uiSchema = {
  uiSchema: {
    str: fieldProps,
    strArr: fieldProps,
    obj: fieldProps,
    objArr: fieldProps,
  },
};

test("Typeahead render", () => {
  const component = renderer.create(
    <Form
      schema={schema}
      uiSchema={uiSchema}
      formData={formData}
      fields={fields}
    />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
