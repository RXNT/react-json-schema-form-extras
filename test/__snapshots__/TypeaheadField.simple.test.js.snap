// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typeahead render 1`] = `
<form
  acceptCharset={undefined}
  action={undefined}
  autoComplete={undefined}
  className="rjsf"
  encType={undefined}
  id={undefined}
  method={undefined}
  name={undefined}
  noValidate={false}
  onSubmit={[Function]}
  target={undefined}
>
  <div
    className="form-group field field-object"
  >
    <fieldset>
      <div
        className="form-group field field-string"
      >
        <label
          className="control-label"
          htmlFor="root_str"
        >
          str
        </label>
        <div />
        <input
          autoFocus={false}
          className="form-control"
          disabled={false}
          id="root_str"
          label="str"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder=""
          readOnly={false}
          required={false}
          type="text"
          value="some"
        />
        <div />
        <div />
      </div>
      <div
        className="form-group field field-array"
      >
        <fieldset
          className="field field-array field-array-of-string"
        >
          <legend
            id="root_strArr__title"
          >
            strArr
          </legend>
          <div
            className="row array-item-list"
          >
            <div
              className="array-item"
            >
              <div
                className="col-xs-9"
              >
                <div
                  className="form-group field field-string"
                >
                  <div />
                  <div />
                  <input
                    autoFocus={false}
                    className="form-control"
                    disabled={false}
                    id="root_strArr_0"
                    label={undefined}
                    onBlur={[Function]}
                    onChange={[Function]}
                    placeholder=""
                    readOnly={false}
                    required={true}
                    type="text"
                    value="some"
                  />
                  <div />
                  <div />
                </div>
              </div>
              <div
                className="col-xs-3 array-item-toolbox"
              >
                <div
                  className="btn-group"
                  style={
                    Object {
                      "display": "flex",
                      "justifyContent": "space-around",
                    }
                  }
                >
                  <button
                    className="btn btn-danger array-item-remove"
                    disabled={false}
                    onClick={[Function]}
                    style={
                      Object {
                        "flex": 1,
                        "fontWeight": "bold",
                        "paddingLeft": 6,
                        "paddingRight": 6,
                      }
                    }
                    tabIndex="-1"
                    type="button"
                  >
                    <i
                      className="glyphicon glyphicon-remove"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
          <div
            className="row"
          >
            <p
              className="col-xs-3 col-xs-offset-9 array-item-add text-right"
            >
              <button
                className="btn btn-info btn-add col-xs-12"
                disabled={false}
                onClick={[Function]}
                tabIndex="0"
                type="button"
              >
                <i
                  className="glyphicon glyphicon-plus"
                />
              </button>
            </p>
          </div>
        </fieldset>
        <div />
        <div />
      </div>
      <div
        className="form-group field field-object"
      >
        <fieldset>
          <legend
            id="root_obj__title"
          >
            obj
          </legend>
          <div
            className="form-group field field-string"
          >
            <label
              className="control-label"
              htmlFor="root_obj_name"
            >
              name
            </label>
            <div />
            <input
              autoFocus={false}
              className="form-control"
              disabled={false}
              id="root_obj_name"
              label="name"
              onBlur={[Function]}
              onChange={[Function]}
              placeholder=""
              readOnly={false}
              required={false}
              type="text"
              value="other"
            />
            <div />
            <div />
          </div>
        </fieldset>
        <div />
        <div />
      </div>
      <div
        className="form-group field field-array"
      >
        <fieldset
          className="field field-array field-array-of-object"
        >
          <legend
            id="root_objArr__title"
          >
            objArr
          </legend>
          <div
            className="row array-item-list"
          >
            <div
              className="array-item"
            >
              <div
                className="col-xs-9"
              >
                <div
                  className="form-group field field-object"
                >
                  <fieldset>
                    <div
                      className="form-group field field-string"
                    >
                      <label
                        className="control-label"
                        htmlFor="root_objArr_0_name"
                      >
                        name
                      </label>
                      <div />
                      <input
                        autoFocus={false}
                        className="form-control"
                        disabled={false}
                        id="root_objArr_0_name"
                        label="name"
                        onBlur={[Function]}
                        onChange={[Function]}
                        placeholder=""
                        readOnly={false}
                        required={false}
                        type="text"
                        value="other"
                      />
                      <div />
                      <div />
                    </div>
                  </fieldset>
                  <div />
                  <div />
                </div>
              </div>
              <div
                className="col-xs-3 array-item-toolbox"
              >
                <div
                  className="btn-group"
                  style={
                    Object {
                      "display": "flex",
                      "justifyContent": "space-around",
                    }
                  }
                >
                  <button
                    className="btn btn-danger array-item-remove"
                    disabled={false}
                    onClick={[Function]}
                    style={
                      Object {
                        "flex": 1,
                        "fontWeight": "bold",
                        "paddingLeft": 6,
                        "paddingRight": 6,
                      }
                    }
                    tabIndex="-1"
                    type="button"
                  >
                    <i
                      className="glyphicon glyphicon-remove"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
          <div
            className="row"
          >
            <p
              className="col-xs-3 col-xs-offset-9 array-item-add text-right"
            >
              <button
                className="btn btn-info btn-add col-xs-12"
                disabled={false}
                onClick={[Function]}
                tabIndex="0"
                type="button"
              >
                <i
                  className="glyphicon glyphicon-plus"
                />
              </button>
            </p>
          </div>
        </fieldset>
        <div />
        <div />
      </div>
    </fieldset>
    <div />
    <div />
  </div>
  <p>
    <button
      className="btn btn-info"
      type="submit"
    >
      Submit
    </button>
  </p>
</form>
`;
